### PR TITLE
chore(sql): add varchar LPad/RPad functions

### PR DESCRIPTION
**Pull request by eugenels**

### CODE REVIEW

**Comment by ideoma** [comment]
#### [PR Coverage check]

:heart_eyes: **pass** : 177 / 202 (87.62%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|io/questdb/griffin/engine/functions/str/RPadVarcharVarcharFunctionFactory.java|56|67|83.58%|
|:large_blue_circle:|io/questdb/griffin/engine/functions/str/LPadVarcharVarcharFunctionFactory.java|55|65|84.62%|
|:large_blue_circle:|io/questdb/griffin/engine/functions/str/LPadVarcharFunctionFactory.java|33|35|94.29%|
|:large_blue_circle:|io/questdb/griffin/engine/functions/str/RPadVarcharFunctionFactory.java|33|35|94.29%|

**Comment by ideoma** [comment]
#### [PR Coverage check]

:heart_eyes: **pass** : 177 / 202 (87.62%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|io/questdb/griffin/engine/functions/str/RPadVarcharVarcharFunctionFactory.java|56|67|83.58%|
|:large_blue_circle:|io/questdb/griffin/engine/functions/str/LPadVarcharVarcharFunctionFactory.java|55|65|84.62%|
|:large_blue_circle:|io/questdb/griffin/engine/functions/str/LPadVarcharFunctionFactory.java|33|35|94.29%|
|:large_blue_circle:|io/questdb/griffin/engine/functions/str/RPadVarcharFunctionFactory.java|33|35|94.29%|

File: core/src/main/java/io/questdb/griffin/engine/functions/str/LPadStrVarcharFunctionFactory.java [hunk 1]
```diff
@@ -0,0 +1,137 @@
...
+import io.questdb.griffin.FunctionFactory;
+import io.questdb.griffin.SqlExecutionContext;
+import io.questdb.griffin.engine.functions.TernaryFunction;
+import io.questdb.griffin.engine.functions.VarcharFunction;
+import io.questdb.std.IntList;
+import io.questdb.std.ObjList;
+import io.questdb.std.str.*;
+import org.jetbrains.annotations.Nullable;
+
+public class LPadStrVarcharFunctionFactory implements FunctionFactory {
```
[the remark this answers is not in the document]
Comment (eugenels) [reply core/src/main/java/io/questdb/griffin/engine/functions/str/LPadStrVarcharFunctionFactory.java]: addressed in: 01d1b717e1e6871cff4e1b9a777a864a943f35e4

File: core/src/main/java/io/questdb/griffin/engine/functions/str/LPadVarcharFunctionFactory.java [hunk 2]
```diff
@@ -0,0 +1,135 @@
...
+        }
+
+        @Override
+        public void getVarchar(Record rec, Utf8Sink utf8Sink) {
+            utf8Sink.put(lPad(strFunc.getVarcharA(rec), lenFunc.getInt(rec), sink));
+        }
+
+        @Override
+        public Utf8Sequence getVarcharB(final Record rec) {
+            return lPad(strFunc.getVarcharA(rec), lenFunc.getInt(rec), sinkB);
```
[the remark this answers is not in the document]
Comment (eugenels) [reply core/src/main/java/io/questdb/griffin/engine/functions/str/LPadVarcharFunctionFactory.java]: addressed in: 01d1b717e1e6871cff4e1b9a777a864a943f35e4

File: core/src/main/java/io/questdb/griffin/engine/functions/str/RPadStrVarcharFunctionFactory.java [hunk 3]
```diff
@@ -0,0 +1,142 @@
...
+                sink.clear();
+
+                final int length = Utf8s.validateUtf8(str);
+                if (len > length) {
+                    sink.put(str);
+                    final int fillTextLen = Utf8s.validateUtf8(fillText);
+                    for (int i = 0, n = (len - length) / fillTextLen; i < n; i++) {
+                        sink.put(fillText);
+                    }
+                    Utf8s.strCpy(fillText, 0, (len - length) % fillTextLen, sink);
```
[the remark this answers is not in the document]
Comment (eugenels) [reply core/src/main/java/io/questdb/griffin/engine/functions/str/RPadStrVarcharFunctionFactory.java]: addressed in: 01d1b717e1e6871cff4e1b9a777a864a943f35e4

File: core/src/test/java/io/questdb/test/griffin/engine/functions/str/LPadRPadFunctionsTest.java [hunk 4]
```diff
@@ -0,0 +1,93 @@
+package io.questdb.test.griffin.engine.functions.str;
+
+import io.questdb.griffin.SqlException;
+import org.junit.Test;
+import io.questdb.test.AbstractCairoTest;
+
+public class LPadRPadFunctionsTest extends AbstractCairoTest {
```
[the remark this answers is not in the document]
Comment (eugenels) [reply core/src/test/java/io/questdb/test/griffin/engine/functions/str/LPadRPadFunctionsTest.java:7]: addressed in: 01d1b717e1e6871cff4e1b9a777a864a943f35e4

File: core/src/main/java/io/questdb/griffin/engine/functions/str/RPadStrVarcharFunctionFactory.java [hunk 5]
```diff
@@ -0,0 +1,142 @@
...
+            ObjList<Function> args,
+            IntList argPositions,
+            CairoConfiguration configuration,
+            SqlExecutionContext sqlExecutionContext
+    ) {
+        final Function strFunc = args.getQuick(0);
+        final Function lenFunc = args.getQuick(1);
+        final Function fillTextFunc = args.getQuick(2);
+        final int maxLength = configuration.getStrFunctionMaxBufferLength();
+        return new RPadVarcharFunc(strFunc, lenFunc, fillTextFunc, maxLength);
```
[the remark this answers is not in the document]
Comment (eugenels) [reply core/src/main/java/io/questdb/griffin/engine/functions/str/RPadStrVarcharFunctionFactory.java]: addressed in: 01d1b717e1e6871cff4e1b9a777a864a943f35e4

**Review by ideoma — APPROVED** [review]

**Comment by ideoma** [comment]
Is it worth updating `LPadStrFunction` to use `getStrB`?

**Comment by ideoma** [comment]
#### [PR Coverage check]

:heart_eyes: **pass** : 177 / 202 (87.62%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|io/questdb/griffin/engine/functions/str/RPadVarcharVarcharFunctionFactory.java|56|67|83.58%|
|:large_blue_circle:|io/questdb/griffin/engine/functions/str/LPadVarcharVarcharFunctionFactory.java|55|65|84.62%|
|:large_blue_circle:|io/questdb/griffin/engine/functions/str/LPadVarcharFunctionFactory.java|33|35|94.29%|
|:large_blue_circle:|io/questdb/griffin/engine/functions/str/RPadVarcharFunctionFactory.java|33|35|94.29%|

**Pull request merged**